Signal/src/ViewControllers/MessagesViewController.m

```diff
@@ -2704,48 +3032,125 @@ - (TSInteraction *)interactionAtIndexPath:(NSIndexPath *)indexPath {
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
 
     // Initiate and prepare the recorder
-    _audioRecorder          = [[AVAudioRecorder alloc] initWithURL:outputFileURL settings:recordSetting error:NULL];
-    _audioRecorder.delegate = self;
-    _audioRecorder.meteringEnabled = YES;
-    [_audioRecorder prepareToRecord];
-}
-
-- (void)audioRecorderDidFinishRecording:(AVAudioRecorder *)recorder successfully:(BOOL)flag {
-    if (flag) {
-        NSData *audioData = [NSData dataWithContentsOfURL:recorder.url];
-        SignalAttachment *attachment =
-            [SignalAttachment attachmentWithData:audioData dataUTI:(NSString *)kUTTypeMPEG4Audio filename:nil];
-        if (!attachment ||
-            [attachment hasError]) {
-            DDLogWarn(@"%@ %s Invalid attachment: %@.",
-                      self.tag,
-                      __PRETTY_FUNCTION__,
-                      attachment ? [attachment errorName] : @"Missing data");
-            [self showErrorAlertForAttachment:attachment];
-        } else {
-            [self tryToSendAttachmentIfApproved:attachment];
-        }
+    self.audioRecorder = [[AVAudioRecorder alloc] initWithURL:fileURL
+                                                     settings:@{
+                                                         AVFormatIDKey : @(kAudioFormatMPEG4AAC),
+                                                         AVSampleRateKey : @(44100),
+                                                         AVNumberOfChannelsKey : @(2),
+                                                         AVEncoderBitRateKey: @(128 * 1024),
+                                                     }
+                                                        error:&error];
+    if (error) {
+        DDLogError(@"%@ Couldn't create audioRecorder: %@", self.tag, error);
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
+
+    self.audioRecorder.meteringEnabled = YES;
+
+    if (![self.audioRecorder prepareToRecord]) {
+        DDLogError(@"%@ audioRecorder couldn't prepareToRecord.", self.tag);
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
+
+    if (![self.audioRecorder record]) {
+        DDLogError(@"%@ audioRecorder couldn't record.", self.tag);
+        [self resetRecordingVoiceMemo];
+        OWSAssert(0);
+        return;
+    }
+}
+
+- (void)endRecordingVoiceMemo
+{
+    OWSAssert([NSThread isMainThread]);
+
+    DDLogInfo(@"endRecordingVoiceMemo");
+
+    if (!self.audioRecorder) {
+        DDLogError(@"%@ Missing audioRecorder", self.tag);
+        OWSAssert(0);
+        return;
+    }
+
+    NSTimeInterval currentTime = self.audioRecorder.currentTime;
+
+    [self.audioRecorder stop];
+
+    const NSTimeInterval kMinimumRecordingTimeSeconds = 1.f;
+    if (currentTime < kMinimumRecordingTimeSeconds) {
+        DDLogInfo(@"Discarding voice memo; too short.");
```

Comment:
Yes, this is the crux of what I want to discuss with you and Riya before going any further. What "crutches"/"hints" and other polish we want to add around voice memoes, if any.